lib/msf/core/post/windows/services.rb

```diff
@@ -145,48 +174,48 @@ def open_service_handle(manager, name, access)
     end
   end
 
+  #
   # Yield each service name on the remote host
   #
-  # @todo Allow operating on a remote host
   # @yield [String] Case-sensitive name of a service
+  #
+  # @todo Allow operating on a remote host
+  #
   def each_service(&block)
-    if load_extapi
-      session.extapi.service.enumerate.each(&block)
-    else
-      serviceskey = "HKLM\\SYSTEM\\CurrentControlSet\\Services"
-
-      keys = registry_enumkeys(serviceskey)
-      keys.each do |sk|
-        srvtype = registry_getvaldata("#{serviceskey}\\#{sk}","Type")
-        # From http://support.microsoft.com/kb/103000
-        #
-        # 0x1            A Kernel device driver.
-        #
-        # 0x2            File system driver, which is also
-        #                a Kernel device driver.
-        #
-        # 0x4            A set of arguments for an adapter.
-        #
-        # 0x10           A Win32 program that can be started
-        #                by the Service Controller and that
-        #                obeys the service control protocol.
-        #                This type of Win32 service runs in
-        #                a process by itself.
-        #
-        # 0x20           A Win32 service that can share a process
-        #                with other Win32 services.
-        if srvtype == 32 || srvtype == 16
-          yield sk
-        end
+    if session.commands.include?(Rex::Post::Meterpreter::Extensions::Stdapi::COMMAND_ID_STDAPI_REGISTRY_ENUM_KEY)
+      begin
+        return session.extapi.service.enumerate.each(&block)
+      rescue Rex::Post::Meterpreter::RequestError => e
+        vprint_error("Request Error #{e} falling back to registry technique")
       end
+    end
 
-      keys
+    serviceskey = "HKLM\\SYSTEM\\CurrentControlSet\\Services"
+
+    keys = registry_enumkeys(serviceskey)
+    keys.each do |sk|
+      service_type = registry_getvaldata("#{serviceskey}\\#{sk}", 'Type').to_s
+      service_type = (service_type.starts_with?('0x') ? service_type.to_i(16) : service_type.to_i)
```

Comment:
There is a potential issue here if the string returned from `registry_getvaldata` is empty as you will end up setting `service_type` to `0` as a result:

```
irb(main):001:0> nil.to_s
=> ""
irb(main):002:0> "".to_i
=> 0
irb(main):003:0> 
```
Thankfully none of the current service values include 0 so this won't impact anything as it stands now but this is something to be aware of.

Probably best to include an extra guard clause here to return if the string is empty.

Reply:
Resolved with https://github.com/rapid7/metasploit-framework/pull/16928/commits/51b6d0b596eb5ff1f9ff7f64e196146ac743eb98